### ISSUE TITLE
[CURATOR-349] Expose extra metrics in TracerDriver

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#869]
<p>Currently, the TracerDriver exposed the latency of ZK operations, in multi-tenant environment, extra metrics are required to help tracing and monitoring:</p>

<ul>
	<li>the bytes being sent and received, so we can monitor the client usage scenarios.</li>
	<li>which ensemble participant the client is talking to, used to find out the problematic Zk server when the issue happened.</li>
	<li>the z-node path, to easily find out which z-node caused the problem, like high load, etc.</li>
</ul>

---
<details><summary><i>Originally reported by <a title="Fangmin Lv" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-349" target="_blank">Expose extra metrics in TracerDriver</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Fangmin Lv" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-10-25T16:18:03+01:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>:</i>
<p>To add this support, the TracerDriver interface need to be changed, if we want to make it backward compatible, maybe a new interface such as AdvancedTracerDriver is a better idea. But I'm doubt this is a real issue we need to consider. Any idea?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>:</i>
<p>Hi guys, any comment on this?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>I don't really have a strong opinion either way as I don't use the TraceDriver currently. But if you'd like to provide a PR with your proposed changes, I'd be happy to review it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>:</i>
<p>Thanks Cameron, I'll provide a PR soon.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user lvfangmin opened a pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/164" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/164</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/lvfangmin/curator" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/lvfangmin/curator</a> <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/curator/pull/164.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/164.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #164</p>

<hr />
<p>commit c3e2f24525cc722642b108e4da95ef3f73373d2c<br/>
Author: allenlyu <allenlyu@fb.com><br/>
Date:   2016-09-20T21:56:47Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/164" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/164</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user lvfangmin opened a pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/lvfangmin/curator" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/lvfangmin/curator</a> <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/curator/pull/165.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #165</p>

<hr />
<p>commit 48c687c241e4b95b4384c91609cc7b4a5dd68ae6<br/>
Author: allenlyu <allenlyu@fb.com><br/>
Date:   2016-09-20T21:56:47Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lvfangmin">lvfangmin</a>:</i>
<p>The <a href="https://github.com/apache/curator/pull/164" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/164</a> pull request was closed by mistake, I've sent a new pull request, sorry for the confusion.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user cammckenzie commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Thanks for the PR,<br/>
    The content looks OK, but I think that the interface changes need to be made backwards compatible. Can we change the OperationTrace to extend the existing TimerTrace?</p>

<p>    @Randgalt I'm not really familiar with this code, so it could probably use another set of eyes.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Thanks @cammckenzie for review, the TimerTrace is actually an internal helper class, it shouldn't have any backwards compatible problem. It is the TracerDriver that may have backwards compatible issue, but more strictly it's not a backwards compatible issue, it's that the users who are using TracerDriver have to change their implementation based on the new interface.</p>

<p>    @Randgalt can you help review on it, it's a feature required internally and we want to have it as soon as possible, thanks <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    ping.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user cammckenzie commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    I've had a bit more of a look and I think that the changes are OK. As you mentioned, the changes are only to internal classes. There is a loss of granularity (The TimeTrace was using nanos while your new implementation is using millis), but I doubt that this is an issue in practice. So, I'm happy with the changes.</p>

<p>    @Randgalt is away at the moment, but I think that he should review it when he returns.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @cammckenzie  Thanks again for reviewing the code, I think the millis is good enough for tracing. I'll wait @Randgalt's review <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    I'm concerned that this change will force a major version bump from us. Can't this be done with the existing classes and a few new methods?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @Randgalt with new methods in the existing classes will also changes the interface the users are using, it has the same side effect as this one, right?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    I haven't looked at the PR in depth. But, having this PR be backward compatible should be explored. Possibly by emulating the old SPI/API.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @Randgalt thanks for giving suggestions, there is another option: we create a new tracer interface, like AdvancedTracerDriver which will expose more metrics like we're doing in this PR. What's your idea?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Thanks  @Randgalt, I'll update the PR based on this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/165#discussion_r82691347" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165#discussion_r82691347</a></p>

<p>    &#8212; Diff: curator-client/src/main/java/org/apache/curator/ConnectionState.java &#8212;<br/>
    @@ -199,13 +202,28 @@ private synchronized void checkTimeouts() throws Exception</p>
 {
     log.error(String.format("Connection timed out for connection string (%s) and timeout (%d) / elapsed (%d)", zooKeeper.getConnectionString(), connectionTimeoutMs, elapsed), connectionLossException);
 }
<ul class="alternate" type="square">
	<li>tracer.get().addCount("connections-timed-out", 1);<br/>
    +    new EventTrace("connections-timed-out", tracer.get(), getSessionId()).commit();<br/>
 throw connectionLossException;<br/>
     }<br/>
 }<br/>
     }<br/>
 }</li>
</ul>


<p>    +    /**<br/>
    +     * Return the current session id<br/>
    +     */<br/>
    +    public long getSessionId() {<br/>
    +long sessionId = -1;<br/>
    +if (isConnected()) {<br/>
    +    try {<br/>
    +sessionId = getZooKeeper().getSessionId();<br/>
    &#8212; End diff &#8211;</p>

<p>    Session id is a useful information to track the client side issue, it's kind of identity, so we would like to include it. I'll change it to use zookeeper.getZookeeper() directly here.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Hi @Randgalt, in case you're not aware, I've updated the code base are you opinion, let me know if you have other comments, thanks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    I think we're almost there. Have you tried merging this with the <a href="https://issues.apache.org/jira/browse/CURATOR-3" title="LeaderLatch race condition causing extra nodes to be added in Zookeeper Edit" class="issue-link" data-issue-key="CURATOR-3"><del>CURATOR-3</del></a>.0 branch yet? Because this is a biggish change it would be good to have a separate PR based off of <a href="https://issues.apache.org/jira/browse/CURATOR-3" title="LeaderLatch race condition causing extra nodes to be added in Zookeeper Edit" class="issue-link" data-issue-key="CURATOR-3"><del>CURATOR-3</del></a>.0 to make sure there are no conflicts there.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Thanks @Randgalt, I've tried to merge this, there is not much conflict, I'll send a separate PR based on <a href="https://issues.apache.org/jira/browse/CURATOR-3" title="LeaderLatch race condition causing extra nodes to be added in Zookeeper Edit" class="issue-link" data-issue-key="CURATOR-3"><del>CURATOR-3</del></a>.0 today. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user lvfangmin opened a pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/169" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/169</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/lvfangmin/curator" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/lvfangmin/curator</a> <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a>-3.0</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/curator/pull/169.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/169.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #169</p>

<hr />
<p>commit 083c6d66b452b9e0ed26e2bc765d3c5da422ae21<br/>
Author: allenlyu <allenlyu@fb.com><br/>
Date:   2016-09-20T21:56:47Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-349" title="Expose extra metrics in TracerDriver" class="issue-link" data-issue-key="CURATOR-349"><del>CURATOR-349</del></a> Expose extra metrics in TracerDriver</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/169" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/169</a></p>

<p>    Rebased on <a href="https://issues.apache.org/jira/browse/CURATOR-3" title="LeaderLatch race condition causing extra nodes to be added in Zookeeper Edit" class="issue-link" data-issue-key="CURATOR-3"><del>CURATOR-3</del></a>.0 branch, same as the code in <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a>.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Hi @Randgalt, I've opened another PR for 3.0 branch, could you please take a look? Let me know if there is any issue.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @Randgalt @cammckenzie, Facebook is going to hosted a Zookeeper Meetup, there will be tech talks, swags, drinks and snacks, please join us if you're interested in: <a href="https://www.facebook.com/events/1228722650504268/" class="external-link" target="_blank" rel="nofollow noopener">https://www.facebook.com/events/1228722650504268/</a>.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user cammckenzie commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Thanks @lvfangmin, unfortunately I'm somewhat geographically challenged living on the other side of the world. Sounds like fun though.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Sounds like they'll have a remote option - I'm remote as well</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Yes, feel free to join remote, we'll post the ways to join online soon.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/169" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/169</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @Randgalt @cammckenzie Thanks for reviewing the code, we're planning to unify the Zookeeper java client to be Curator inside Facebook, so it's really important for us to track the client side metrics. </p>

<p>    Since we're depending on this patch to unify the java client, may I ask what's the next scheduled release? And will this patch be in it?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @lvfangmin </p>

<p>    > we're planning to unify the Zookeeper java client to be Curator inside Facebook</p>

<p>    Wow - cool! Can I publicize that on Curator's wiki?</p>

<p>    > may I ask what's the next scheduled release? And will this patch be in it?</p>

<p>    Personally, I'm waiting on <a href="https://issues.apache.org/jira/browse/ZOOKEEPER-2169" title="Enable creation of nodes with TTLs" class="issue-link" data-issue-key="ZOOKEEPER-2169"><del>ZOOKEEPER-2169</del></a> and <a href="https://issues.apache.org/jira/browse/ZOOKEEPER-1525" title="Plumb ZooKeeperServer object into auth plugins" class="issue-link" data-issue-key="ZOOKEEPER-1525"><del>ZOOKEEPER-1525</del></a> and was planning on doing a build when those are ready. Supposedly it will be in ZK 3.6. How long can you wait?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    @Randgalt we haven't started the unifying progress yet, since we were blocked by this feature, it would be better to publicize it on Curator's wiki when we're actually using it inside FB, I'll let you know, and feel free to publicize this on Curator's wiki then <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>    Unifying the ZooKeeper Java client is our Q4 Goal, and it takes time to ask/help the internal teams to switch to Curator, so it would be great if we can have a release at the end of Oct or beginning of Nov.  Inside Facebook, we're still using ZK 3.4, so can we have a release in Curator 2.xxx which include this change?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Let's see how ZK goes in the next week or so. If there's no movement, we can release Curator. We tend to move a lot faster than them anyway.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/165</a></p>

<p>    Cool, thanks @Randgalt, let me know when we decided the released schedule.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/172#discussion_r86045533" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172#discussion_r86045533</a></p>

<p>    &#8212; Diff: curator-client/src/main/java/org/apache/curator/ConnectionState.java &#8212;<br/>
    @@ -213,16 +213,14 @@ private synchronized void checkTimeouts() throws Exception</p>
<ul>
	<li>Return the current session id<br/>
  */<br/>
 public long getSessionId() {</li>
</ul>
<ul class="alternate" type="square">
	<li>long sessionId = -1;</li>
	<li>if (isConnected()) {</li>
	<li>try {</li>
	<li>ZooKeeper zk = zooKeeper.getZooKeeper();</li>
	<li>if (zk != null) 
{
    -    sessionId = zk.getSessionId();
    -}</li>
	<li>} catch (Exception e) {</li>
	<li>// Ignore the exception<br/>
    +long sessionId = 0;
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Why did we lose the `isConnected()` check?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    (Not related with this diff) <br/>
    @Randgalt , do we have the schedule of Curator release, are we still blocked by <a href="https://issues.apache.org/jira/browse/ZOOKEEPER-1525" title="Plumb ZooKeeperServer object into auth plugins" class="issue-link" data-issue-key="ZOOKEEPER-1525"><del>ZOOKEEPER-1525</del></a>?</p>

<p>    The <a href="https://issues.apache.org/jira/browse/ZOOKEEPER-1525" title="Plumb ZooKeeperServer object into auth plugins" class="issue-link" data-issue-key="ZOOKEEPER-1525"><del>ZOOKEEPER-1525</del></a> looks good to me except the new additional parameters concern, if we can refactor that a bit, I'll add a +1 and ask Ben (one of the ZooKeeper committer, also my team member) to accelerate the review process and accept it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    Btw, I also like to get this in the next release <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/172#discussion_r87313550" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172#discussion_r87313550</a></p>

<p>    &#8212; Diff: curator-client/src/main/java/org/apache/curator/ConnectionState.java &#8212;<br/>
    @@ -213,16 +213,14 @@ private synchronized void checkTimeouts() throws Exception</p>
<ul>
	<li>Return the current session id<br/>
  */<br/>
 public long getSessionId() {</li>
</ul>
<ul class="alternate" type="square">
	<li>long sessionId = -1;</li>
	<li>if (isConnected()) {</li>
	<li>try {</li>
	<li>ZooKeeper zk = zooKeeper.getZooKeeper();</li>
	<li>if (zk != null) 
{
    -    sessionId = zk.getSessionId();
    -}</li>
	<li>} catch (Exception e) {</li>
	<li>// Ignore the exception<br/>
    +long sessionId = 0;
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Yes, I think you're right. I think I can accept this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    I'll start working on a release</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    Thanks @Randgalt, please let me know when the new release is published 👍 </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    Hi @Randgalt, hope you have a great weekend <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> <br/>
    Sorry about this, but could you please give me some updates about the release, so I can plan this weeks' work about whether I can start migrating the internal projects to Curator or not.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    @lvfangmin do you subscribe to dev@curator? We released this morning and an email was sent to that list. So, the bits are available now.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user lvfangmin commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/172</a></p>

<p>    Ah, just saw it, thanks @Randgalt!</p>
